### ISSUE TITLE
A workaround for GSL qag error in ps.c

### ISSUE THREAD

**Issue by SuvedhaNaik** [BradGreig/21CMMC#6]
Hello All, 

I'm trying to constrain an inflation model with a modified power spectrum template using 21CMMC. 
The free parameters of my model are the amplitude of the primordial power spectra, A at scale k_p.
When A is 10^{-8}, I get `GSL qag error (roundoff error prevents tolerance from being achieved)`. 
I tracked the source of this error, and it is coming from two functions in `ps.c`:
(i)  `dsigmasqdm_z0()`
(ii)  `sigma_z0()`
By changing the relative tolerance in (i) from 10^(-6) to 10^(-4) and (ii) from 10^(-6) to 10^(-5), 
I can make the code run for A=10^(-9). 

My questions are:
a. Is there a way to achieve this without changing the relative tolerance?
b. Is it reasonable to increase the relative tolerance as mentioned above, particularly for the functions (i) and (ii)?
In other words, am I losing any important information by making these approximations?

Any comments would be helpful.

Thanks in advance,
Suvedha

**Comment by BradGreig** [comment]
Hi @SuvedhaNaik, I guess for a definitive answer it is hard to say as it really depends on the function you are attempting to integrate and its general behaves over the desired parameter space.

Naively I do not think that lowering the relative tolerance should cause too much of an issue. You could always try comparing the result obtained from lowering the tolerance to that achieved through another method, for example coding it up in Python to double check the performance. If the two converge, then the lower tolerance would be sufficient.

As to alternatives to changing the tolerance, I guess you could look into alternative integration routines. GSL has a variety of integration routines, some of which might be more suitable for the exact integral you are aiming to solve.

Sorry I couldn't be more definitive with my response. Let me know if you have further questions and I'll try and assist as best as I can.

**Comment by BradGreig** [comment]
Also, in case you are interested and/or not aware, we are actively working on newer versions of [21cmFAST](https://github.com/21cmfast/21cmFAST) and [21CMMC](https://github.com/21cmfast/21CMMC). Though you'd still encounter the same issue under these newer versions.

**Comment by SuvedhaNaik** [comment]
Dear @BradGreig,

Thank you very much for your reply. I'm exploring how the relative tolerance would affect the integrations
and different types of integrations in GSL.  

Regarding the newer versions of 21CMMC, I had some difficulties in editing the source code. 
I don't have much knowledge on editing source code from pip installed packages and unfortunately could not spend much time on it. 
So, I switched to C versions.  

Thanks again for your comments.
Regards,
Suvedha

**Comment by BradGreig** [comment]
Hi @SuvedhaNaik, no problems. Please let me know if you have any further questions/problems.
